medschool/src/parse.rs

```diff
@@ -200,43 +203,53 @@ pub fn parse_docs_into_set<'a>(
 /// DFS helper function to resolve the references of the types. Returns the docs of the fields of
 /// the field we're currently looking at search till its leaf nodes. The leaf here means a primitive
 /// type for which we don't have a [`PartialType`].
+#[tracing::instrument(level = "trace", ret)]
 fn dfs_fields<'a, const MAX_RECURSION_LEVEL: usize>(
     field: &PartialField<'a>,
     types: &'a HashSet<PartialType>,
     cache: &mut HashMap<&'a str, Vec<String>>,
     recursion_level: &mut usize,
-) -> Vec<String> {
+    schema: &Option<JSONSchema>,
+) -> Result<Vec<String>, DocsError> {
     if *recursion_level >= MAX_RECURSION_LEVEL {
-        return vec!["Recursion limit reached".to_string()];
+        return Ok(vec!["Recursion limit reached".to_string()]);
+    }
+    if let Some(schema) = schema {
+        validate_json_blocks(&field.docs, schema)
+            .inspect_err(|e| error!(error = ?e, context=?field, "Error validating JSON blocks"))?;
+    }
+    // get the type of the field from the types set to recurse into it's fields
+    match types.get(&field.ty) {
```

Comment:
I don't think you need to change this so much just due to the function returning a `Result` now?

The logic inside this `match` looks very similar to the iterator (at a glance at least).